File: python/union_find.py
# -*- coding: utf-8 -*-
"""
蟻本準拠のUnionFind
"""

class UnionFind:
    def __init__(self, size):
        self.rank = [0] * size
        self.par = range(size)
        self.g_num = size

    def find(self, x):
        if x == self.par[x]: return x
        self.par[x] = self.find(self.par[x])
        return self.par[x]

    def same(self, x, y):
        return self.find(x) == self.find(y)

    def unite(self, x, y):
        x, y = self.find(x), self.find(y)
        if x == y: return

        self.g_num -= 1
        if (self.rank[x] > self.rank[y]):
            self.par[y] = x
        else:
            self.par[x] = y
            if (self.rank[x] == self.rank[y]): self.rank[y] += 1

    def group_num(self):
        return self.g_num

File: cpp/union_find.cpp
/**
   あり本準拠のUnion-Find

   Verified: AOJ DSL_1_A
**/
struct UnionFind {
  int g_num;
  vector<int> rank, par;
  UnionFind(int size) {
    rank.resize(size, 0); par.resize(size, 0);
    for (int i = 0; i < size; i++) par[i] = i, rank[i] = 0;
    g_num = size;
  }
  bool same(int x, int y) {
    return find(x) == find(y);
  }
  void unite(int x, int y) {
    x = find(x); y = find(y);
    if (x == y) return;
    g_num--;
    if (rank[x] > rank[y]) {
      par[y] = x;
    } else {
      par[x] = y;
      if (rank[x] == rank[y]) rank[y]++;
    }
  }
  int find(int x) {
    if (x == par[x]) return x;
    return par[x] = find(par[x]);
  }
  int group_num() {
    return g_num;
  }
};
